src/ProjectOrigin.WalletSystem.Server/Repositories/CertificateRepository.cs

```diff
@@ -423,50 +423,41 @@ public async Task SetWalletSliceState(Guid sliceId, WalletSliceState state)
     public async Task<IList<WalletSlice>> ReserveQuantity(string owner, string registryName, Guid certificateId,
         uint reserveQuantity)
     {
-        var availableSlices = await GetOwnersAvailableSlices(registryName, certificateId, owner);
-        if (availableSlices.IsEmpty())
-            throw new InvalidOperationException($"Owner has no available slices to reserve");
+        var slices = await GetOwnersAvailableSlices(registryName, certificateId, owner);
+        var willBeAvailable = await GetRegisteringAndAvailableQuantity(registryName, certificateId, owner);
 
-        if (availableSlices.Sum(slice => slice.Quantity) < reserveQuantity)
+        if (slices.Sum(x => x.Quantity) >= reserveQuantity)
         {
-            var registeringAvailableQuantity =
-                await GetRegisteringAndAvailableQuantity(registryName, certificateId, owner);
-            if (registeringAvailableQuantity >= reserveQuantity)
-                throw new TransientException($"Owner has enough quantity, but it is not yet available to reserve");
-            else
-                throw new InvalidOperationException($"Owner has less to reserve than available");
-        }
+            var sumSlicesTaken = 0L;
+            var takenSlices = slices.Where(s => s.State == WalletSliceState.Available)
+                .OrderBy(slice => slice.Quantity)
+                .TakeWhile(slice =>
+                {
+                    var needsMore = sumSlicesTaken < reserveQuantity;
+                    sumSlicesTaken += slice.Quantity;
+                    return needsMore;
+                })
+                .ToList();
 
-        var sumSlicesTaken = 0L;
-        var takenSlices = availableSlices
-            .OrderBy(slice => slice.Quantity)
-            .TakeWhile(slice =>
+            foreach (var slice in takenSlices)
             {
-                var needsMore = sumSlicesTaken < reserveQuantity;
-                sumSlicesTaken += slice.Quantity;
-                return needsMore;
-            })
-            .ToList();
+                await SetWalletSliceState(slice.Id, WalletSliceState.Reserved);
```

Comment:
Det var vidst fordi jeg sad på en forkert branch da jeg tænkte eksemplet, vi bør nok ikke fjerne den guard her vi lavede sidst..

```suggestion
                var rowsChanged = await _connection.ExecuteAsync(
                @"UPDATE wallet_slices
                SET state = @state
                WHERE id = @sliceId
                    AND state = @expected",
                new
                {
                    sliceid = slice.Id,
                    state = WalletSliceState.Reserved,
                    expected = WalletSliceState.Available
                });

            if (rowsChanged != 1)
                throw new InvalidOperationException($"Slice with id {slice.Id} could not be found or was no longer available");
```

Reply:
Done